File: code/ch10/10.2.5-house-rob-II.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        prev = 0
        curr = 0

        for i in range(len(nums) - 1):
            temp = curr
            curr = max(curr, nums[i] + prev)
            prev = temp

        res = curr

        prev = 0
        curr = 0

        for i in range(1, len(nums)):
            temp = curr
            curr = max(curr, nums[i] + prev)
            prev = temp

        return max(res, curr)

File: code/ch10/10.2.5-house-rob-II.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1) return nums[0];

        int prev = 0, curr = 0;

        for(int i = 0; i < nums.length - 1; i++) {
            int tmp = curr;
            curr = Math.max(curr, nums[i] + prev);
            prev = tmp;
        }

        int res = curr;

        prev = 0;
        curr = 0;

        for(int i = 1; i < nums.length; i++) {
            int tmp = curr;
            curr = Math.max(curr, nums[i] + prev);
            prev = tmp;
        }

        return Math.max(res, curr);
    }
}